internal/grpcserver/storage/elastic.go

```diff
@@ -0,0 +1,213 @@
+		e.client.Search.WithTrackTotalHits(true),
+		e.client.Search.WithPretty(),
+	)
+	if err != nil {
+		return 0, err
+	}
+	var r map[string]interface{}
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		return 0, err
+	}
+	id := r["hits"].(map[string]interface{})["total"].(map[string]interface{})["value"].(float64)
+
+	return uint32(id), nil
+}
+
+func (e *ElasticStorage) GetAll() ([]models.User, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	var buf bytes.Buffer
+	res, err := e.client.Search(
+		e.client.Search.WithContext(context.Background()),
+		e.client.Search.WithIndex("user"),
+		e.client.Search.WithBody(&buf),
+		e.client.Search.WithTrackTotalHits(true),
+		e.client.Search.WithPretty(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var r map[string]interface{}
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		return nil, err
+	}
+
+	var users []models.User
+	for _, hit := range r["hits"].(map[string]interface{})["hits"].([]interface{}) {
+		id, err := strconv.Atoi(hit.(map[string]interface{})["_id"].(string))
+		if err != nil {
+			return nil, err
+		}
+		data := hit.(map[string]interface{})["_source"]
+		name := data.(map[string]interface{})["name"].(string)
+		email := data.(map[string]interface{})["email"].(string)
+		age := data.(map[string]interface{})["age"].(float64)
+
+		user := models.User{ID: uint32(id), Name: name, Email: email, Age: uint8(age)}
+		users = append(users, user)
+	}
+
+	return users, nil
+}
+
+func (e *ElasticStorage) Retrieve(id uint32) (models.User, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	var buf bytes.Buffer
+	query := map[string]interface{}{
+		"query": map[string]interface{}{
+			"match": map[string]interface{}{
+				"_id": strconv.Itoa(int(id)),
+			},
+		},
+	}
+	if err := json.NewEncoder(&buf).Encode(query); err != nil {
+		return models.User{}, err
+	}
+
+	res, err := e.client.Search(
+		e.client.Search.WithContext(context.Background()),
+		e.client.Search.WithIndex("user"),
+		e.client.Search.WithBody(&buf),
+		e.client.Search.WithTrackTotalHits(true),
+		e.client.Search.WithPretty(),
+	)
+	if err != nil {
+		return models.User{}, err
+	}
+	var r map[string]interface{}
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		return models.User{}, err
+	}
+
+	var user models.User
+	for _, hit := range r["hits"].(map[string]interface{})["hits"].([]interface{}) {
+		id, err := strconv.Atoi(hit.(map[string]interface{})["_id"].(string))
+		if err != nil {
+			return models.User{}, err
+		}
+		data := hit.(map[string]interface{})["_source"]
+		name := data.(map[string]interface{})["name"].(string)
+		email := data.(map[string]interface{})["email"].(string)
+		age := data.(map[string]interface{})["age"].(float64)
+
+		user = models.User{ID: uint32(id), Name: name, Email: email, Age: uint8(age)}
+	}
+
+	return user, nil
+}
+
+func (e *ElasticStorage) Add(name, email string, age uint8) (models.User, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
```

Comment:
move `e.lastID` to some variable and release the lock ASAP.

```golang
e.mu.Lock()
e.lastID++
e.mu.Unlock()
var lastID = e.lastID
```
and then use the lastID var